WIP added image-index methods

ep0ll opened the pull request:
### CNB Pack Manifest Commands: Streamlining Multi-Architecture Image Management
This proposal introduces a new **experimental** set of commands to the CNB tooling ecosystem: _**pack manifest**_. These commands empower buildpack authors to efficiently manage and distribute multi-architecture images, aligning with functionalities offered by existing tools like `podman manifest` and `docker manifest`.

### Key Features:

- **Create manifest lists**: Construct local manifest lists for annotating and pushing to container registries.
- **Annotate manifests**: Add essential information like architecture, OS, or variant to existing manifest lists.
- **Add/remove images**: Manage individual image entries within manifest lists.
- **Push manifests**: Publish constructed manifest lists to container registries.
- **Inspect manifests**: View detailed information about locally stored manifest lists.
### Benefits:

**Simplified workflows**: Centralized management of multi-architecture image variants for efficient distribution.
Enhanced interoperability: Potential adherence to established standards for seamless integration with existing tools and runtimes.
**Improved developer experience**: Streamlined workflows and user-friendly commands for managing manifest lists.
**Target Users**:

Buildpack authors seeking to distribute their buildpacks across diverse architectures.
### Example Workflow:

**Create a manifest list**:

```sh
pack manifest create cnbs/sample-package:hello-multiarch-universe \
    cnbs/sample-package:hello-universe \
    cnbs/sample-package:hello-universe-windows
```
**Annotate architecture**:

```sh
pack manifest annotate --arch amd64 cnbs/sample-package:hello-multiarch-universe cnbs/sample-package:hello-universe
pack manifest annotate --arch amd64 cnbs/sample-package:hello-multiarch-universe cnbs/sample-package:hello-universe-windows
```
**Push to registry**:

```sh
pack manifest push cnbs/sample-package:hello-multiarch-universe
```
**Current Status**:

These commands are initially **experimental** and gated behind the pack config experimental flag. They will transition to supported status upon deemed readiness by maintainers.

This innovative addition to the CNB ecosystem empowers buildpack authors to streamline multi-architecture image management, fostering efficient distribution and enhanced developer experiences.

### Development setup:
This guide outlines the steps to set up your development environment and test the pack manifest commands under development in the image-index branch of the `github.com/WYGIN/buildpacks-pack` repository against the `github.com/WYGIN/buildpacks-imgutil` repository.

#### Prerequisites:
- Docker installed and running on your system.
- Git version control system installed.

#### Steps:

1. **Clone the repositories**:

```sh
git clone https://github.com/WYGIN/buildpacks-pack.git
```

2. **Navigate to the buildpacks-pack repository**:
```sh
cd buildpacks-pack
```

3. **Checkout the `image-index` branch**:
```sh
git checkout image-index
```

4. **Build the pack binary**:
```sh
make build
```

5. **Install the pack binary**:
```sh
sudo make install
```

6. **Enable pack's `experimental features`**:
```sh
pack config experimental true
```

7. **Run `pack manifest`**:

```sh
pack manifest create cnbs/sample-package:hello-multiarch-universe \
     cnbs/sample-package:hello-universe \

     cnbs/sample-package:hello-universe-windows
```

```sh
# annotate first image
pack manifest annotate --arch amd64 cnbs/sample-package:hello-multiarch-universe cnbs/sample-package:hello-universe
# annotate second image
pack manifest annotate --arch amd64 cnbs/sample-package:hello-multiarch-universe cnbs/sample-package:hello-universe-windows
```

```sh
pack manifest add cnbs/sample-package:hello-multiarch-universe \
					cnbs/sample-package:hello-universe-riscv-linux
```

```sh
pack manifest delete cnbs/sample-package:hello-multiarch-universe
```

```sh
pack manifest rm cnbs/sample-package:hello-multiarch-universe \
					cnbs/sample-package:hello-universe-windows
```

```sh
pack manifest push cnbs/sample-package:hello-multiarch-universe
```

```sh
pack manifest inspect cnbs/sample-builder:multiarch
```

### Resolves:

1. issue: [Create an new interface to handle Manifest Lists #188](https://github.com/buildpacks/imgutil/issues/188)

ep0ll committed:
WIP added image-index methods

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
fWIP fix lint and run format

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
removed ,gitpod.yml file

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP removed ImageIndexHandler

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix some missing logic

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added registry and insecure fields to IndexAddOptions

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added empty test cases

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added tests but failing

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP removed oci layout test files

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP removed cnb dir from local

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP refactor code to improve readability

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added fake index with empty test cases

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added few tests for fake index

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fixed fake index tests

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP defined errors at one place

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP trying to fix  error when saving index

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix: imgutil#Save() and annotations bugs

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix: all bugs fixed except image hash change when adding image to index

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added tests for all indexes

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP made index.NewIndex to return imgutil.ImageIndex instead of imgutil.Index

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP improved test coverage

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP improved test coverage by adding #Save

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix failing tests for #Add

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix all bugs except mutated configFile and Manifest of image when loaded from local filesystem

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix bug except #Save urls

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix all bugs except #Save URLs

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix: SetURLs for #Save

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP bug fixes

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added goroutines to improve performance

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP minor bug fixes

Signed-off-by: WYGIN <wygininc@gmail.com>

jjbustamante commented on index.go:
```diff
@@ -1,47 +1,2085 @@
+		return err
+	}
+
+	return errors.New(string(mfestBytes))
```
I think the signature for this method is wrong, if we want to return the bytes. probably should return (string, error). 

the output in pack is really weird.

<img width="1482" alt="Screenshot 2024-02-09 at 9 36 03 AM" src="https://github.com/husni-faiz/imgutil/assets/1181799/d81ec65d-aa1e-449b-9863-f7c552517c82">

look it is being printed as **ERROR**, because of that it is not easy to pipe to other tooling like `jq`

ep0ll commented on index.go:
```diff
@@ -1,47 +1,2085 @@
+	Delete() error
+}
+
+var (
+	ErrOSUndefined                        = errors.New("os is undefined")
+	ErrArchUndefined                      = errors.New("architecture is undefined")
+	ErrVariantUndefined                   = errors.New("variant is undefined")
+	ErrOSVersionUndefined                 = errors.New("osVersion is undefined")
+	ErrFeaturesUndefined                  = errors.New("features are undefined")
+	ErrOSFeaturesUndefined                = errors.New("os-features are undefined")
+	ErrURLsUndefined                      = errors.New("urls are undefined")
+	ErrAnnotationsUndefined               = errors.New("annotations are undefined")
+	ErrNoImageOrIndexFoundWithGivenDigest = errors.New("no image/index found with the given digest")
+	ErrConfigFilePlatformUndefined        = errors.New("platform is undefined in config file")
+	ErrManifestUndefined                  = errors.New("manifest is undefined")
+	ErrPlatformUndefined                  = errors.New("platform is undefined")
+	ErrInvalidPlatform                    = errors.New("invalid platform is provided")
+	ErrConfigFileUndefined                = errors.New("config file is undefined")
+	ErrIndexNeedToBeSaved                 = errors.New("image index should need to be saved to perform this operation")
+	ErrUnknownMediaType                   = errors.New("media type not supported")
+	ErrNoImageFoundWithGivenPlatform      = errors.New("no image found with the given platform")
+)
```
@jjbustamante it would be great if you also suggest some changes to error messages

jjbustamante commented on index.go:
```diff
@@ -1,47 +1,2085 @@
+	return path.AppendImage(img, layoutOps...)
+}
+
+func (i *Index) Save() error {
```
I am confused about how this method is working. After I tested the `manifest create` method I can see in my PACK_HOME that the whole Image was saved there in OCI layout format. 


<img width="1792" alt="Screenshot 2024-02-09 at 10 30 14 AM" src="https://github.com/husni-faiz/imgutil/assets/1181799/40bf185a-d70c-4278-a3d8-929f17a05472">

Why do we need the whole image? Image I am creating an Index for images that are X GB in size, are we going to write on disk all the blobs for those images ?

ep0ll replied:
yes, it will write all the images and their blobs, manifests, configs, layers, of an image as well as for index, using [layout#AppendManifest](https://github.com/google/go-containerregistry/blob/8dadbe76ff8c20d0e509406f04b7eade43baa6c1/pkg/v1/layout/write.go#L81) will only append image manifest instead of whole image which will in fact performant, but will not modify the annotated fields of underlying image when image index is annotated(same as other tools like docker and podman), 

since some of the existing buildpacks don't have the `os`, `arch` fields so, when dealing with config file of those images, still those `os` and `arch` will not get updated

as [some existing code](https://github.com/buildpacks/imgutil/blob/f8d38e1de03d30a169908d8b8a4c780a6081f23a/remote/remote.go#L45) is depending on these changes to be reflected, in further phases it might cause annoying bugs i think

so i was also thinking about adding an manifestOnly option to `pack config` just like `pack config experimental < true | false >`, on setting it up it will only deal with image index and not underlying images, thus not storing any images and blobs :)

@jjbustamante what do you think about it, should i refactor the entire code to only generate index manifest only? or should i add make them both work together?

ep0ll committed:
WIP added manifestOnly handler

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP minor changes to ImageIndex#Push

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix: use latest index when pushing index

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added TaggableIndex to push IndexManifest only

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP fix: added decribable methods to TaggableIndex

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added documentation, fix: IndexHandler#Push iissue

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP added tests for index.NewIndex

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
Merge branch 'dev-image-index' into image-index

Signed-off-by: Sai Kiran Maggidi <107541780+WYGIN@users.noreply.github.com>

ep0ll committed:
WIP formatted code after Merge upstream branch

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP: fix bugs

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
fix: tests making 'pack manifest' fail

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
fix: return error when setters called on unknown digest of fake index

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: improved error messages

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
fix: bug causing local images to fail annotate

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
fix: manifest handler not saved annotated images

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: removed IndexHandler

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: remove dead code

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: improve ManifestHandler#Save readability

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: nit all methods except #Add

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: nit nil pointer handling

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: ManifestHandler tests

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
feat: added #Features #OSFeatures #URLs #Annotations methods to imgutil.Image

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: remove IndexHandler

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
fix: removed duplicate stringSlices for features, osFeatures, urls

Signed-off-by: WYGIN <wygininc@gmail.com>

jjbustamante commented on cnb_image.go:
```diff
@@ -496,24 +608,24 @@ func (i *CNBImageCore) SetCreatedAtAndHistory() error {
 }
 
-func getConfigFile(image v1.Image) (*v1.ConfigFile, error) {
+func GetConfigFile(image v1.Image) (*v1.ConfigFile, error) {
```
Why are we making the method Public?

WYGIN replied:
reverting back changes :)

ep0ll replied:
I wanted to make some changes in `local` and `remote` packages using these functions to avoid duplication, but now there is no need of these to be public, making these functions to private

jjbustamante commented on cnb_image.go:
```diff
@@ -496,24 +608,24 @@ func (i *CNBImageCore) SetCreatedAtAndHistory() error {
 }
 
-func getManifest(image v1.Image) (*v1.Manifest, error) {
+func GetManifest(image v1.Image) (*v1.Manifest, error) {
```
Similar to my previous question, why are we making this one public?

ep0ll committed:
refactor: change imgutil.Platform to v1.Platform

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
refactor: local imaages added to index should implement EdditableImage

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
fix: add local images return index not found error

Signed-off-by: WYGIN <wygininc@gmail.com>

ep0ll committed:
WIP: added tests for imgutil#MutateManifest

Signed-off-by: WYGIN <wygininc@gmail.com>

SaikiranIndia committed:
WIP: added tests

Signed-off-by: WYGIN <wygininc@gmail.com>

Author:    WYGIN <wygininc@gmail.com>
Date:      Sat Mar 30 07:52:55 2024 +0000
Changes to be committed:
	modified:   index.go
	modified:   index/new.go
	modified:   index_test.go
	modified:   new.go
	modified:   new_test.go
	modified:   util.go
	modified:   util_test.go

SaikiranIndia committed:
fix: test

Signed-off-by: WYGIN <wygininc@gmail.com>

Author:    WYGIN <wygininc@gmail.com>
Date:      Sat Mar 30 08:07:30 2024 +0000
Changes to be committed:
	modified:   util_test.go

SaikiranIndia committed:
WIP: refactor: improve code readability

Signed-off-by: Sai Kiran <sai90100msk@gmail.com>

jjbustamante commented on cnb_image.go:
```diff
@@ -22,10 +23,13 @@ type CNBImageCore struct {
+	preferredMediaTypes          MediaTypes
+	preserveHistory              bool
+	previousImage                v1.Image
+	os, arch, variant, osVersion string
+	features, osFeatures, urls   []string
+	annotations                  map[string]string
```
I noticed you want to keep a reference to this values to speed up some methods and avoid reading the Config file, but is it really making things faster?, I think the information we try to keep i `CNBImage` is the one we don't have in the underlying `v1.Image`

ep0ll replied:
I felt like there were two advantages to keep `os, arch, variant, osVersion, features, osFeatures, URLs, annotations`, 

1. we don't need to read v1.ConfigFile and v1.Manifest and each Get Operation
2. we can avoid mutating v1.ConfigFile and v1.Manifest and each Set Operation and thus by only mutating the image with all the required changes once in `Image#Save` method

I felt like Mutating v1.Image is a bit Intensive operation and I wanted to avoid that

jjbustamante commented on layout/save.go:
```diff
@@ -27,6 +28,52 @@ func (i *Image) SaveAs(name string, additionalNames ...string) error {
 		ops = append(ops, WithoutLayers())
 	}
 
+	i.Image, err = imgutil.MutateManifest(i, func(mfest *v1.Manifest) {
```
This is why I have some doubts about holding attributes, if we just mutate the config file during the set operations do we need to this?

ep0ll replied:
Updating the v1.ConfigFile doesn't reflect an update in v1.Manifest of v1.Image, we need to update both the v1.ConfigFile and v1.Manifest on each set operation to keep Image Configuration files in sync,
In addition [v1.ConfigFile](https://github.com/google/go-containerregistry/blob/8b3c3036d612bcb3c1147fe11c2d1818dc432329/pkg/v1/config.go#L29-L42) has some missing fields like `Features`, `Annotations`, `URLs` that are only supported in [v1.Manifest's](https://github.com/google/go-containerregistry/blob/8b3c3036d612bcb3c1147fe11c2d1818dc432329/pkg/v1/manifest.go#L25-L53) [Platform](https://github.com/google/go-containerregistry/blob/8b3c3036d612bcb3c1147fe11c2d1818dc432329/pkg/v1/platform.go#L24-L31) field
